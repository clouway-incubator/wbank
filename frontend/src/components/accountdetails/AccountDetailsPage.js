import React from 'react';
import {Link} from 'react-router-dom';
import AccountDetailsTable from './AccountDetailsTable';

export default class AccountDetailsPage extends React.Component {
  constructor(props){
    super(props);

    this.onClickDeleteAccount = this.onClickDeleteAccount.bind(this);
    this.onDepositClick = this.onDepositClick.bind(this);
    this.onWithdrawClick = this.onWithdrawClick.bind(this);
    this.onClickAccountHistory = this.onClickAccountHistory.bind(this);
    this.accountFilter = this.accountFilter.bind(this);
  }

  render() {
    return (
      <div className="jumbotron">
        <h2>Account Details</h2>
        {this.accountFilter()}
        <button className="btn btn-default btn-md" onClick={this.onDepositClick}>Deposit</button>
        <button className="btn btn-default btn-md" onClick={this.onWithdrawClick}>Withdraw</button>     
        <button className="btn btn-default btn-md" onClick={this.onClickDeleteAccount}>Delete</button>
        <button className="btn btn-default btn-md" onClick={this.onClickAccountHistory}>Transactions History</button>
        <div className='div--back-button'>
          <Link to='/accounts'>
            <button className="btn btn-default btn-md">Back</button>
          </Link>
        </div>
      </div>
    );
  }
  
  accountFilter() {
    let targetAccount = undefined;
    let targetId = this.props.match.params.id;
    this.props.accounts.forEach(function(account) {
      if (account.AccountID === targetId) {
        targetAccount = account;
        console.log('accountFilter:', account.AccountID, account.Currency);
      }
    });
    if (targetAccount) {
      return <AccountDetailsTable account={targetAccount} />
    } else {
      return <p>Bad Account ID. Please go back.</p>
    }
  };

  onClickAccountHistory() {
    this.props.history.push('/transactions/' + this.props.match.params.id);
  }

  onClickDeleteAccount() {
    this.props.deleteAccount(this.props.match.params.id).then(() => {
      this.props.history.push('/accounts');
    });
  }

  onDepositClick () {
    let promptInput = prompt(`Please enter the amount you want to deposit:`);
    this.isNumber(promptInput) && promptInput != null && promptInput > 0 
      ? 
      this.props.depositAccount(this.props.match.params.id, parseFloat(promptInput))
      : 
      this.props.showStatusMessage(`Invalid amount. Please, type a valid number to deposit.`);
  }

  onWithdrawClick () {
    let promptInput = prompt(`Please enter the amount you want to withdraw:`);
    this.isNumber(promptInput) && promptInput != null && promptInput > 0 
      ? 
      this.props.withdrawAccount(this.props.match.params.id, parseFloat(promptInput))
      : 
      this.props.showStatusMessage(`Invalid amount. Please, type a valid number to withdraw.`);
  }

  isNumber(n) {
    return !isNaN(parseFloat(n)) && isFinite(n);
  }

}