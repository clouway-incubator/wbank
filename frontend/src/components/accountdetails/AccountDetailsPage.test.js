import React from 'react'
import renderer from 'react-test-renderer'
import { shallow, mount, render } from 'enzyme'
import AccountDetailsPage from './AccountDetailsPage'

describe('AccountDetailsPage', () => {
  let page
  let deleteAccount
  let loadAccountDetails
  let Link
 
  beforeEach(() => {
    deleteAccount = jest.fn()  
    loadAccountDetails = jest.fn()
  })

  it('renders the account details properly', () => {   
    const history = { push: jest.fn() }     
    const tree = renderer.create(
      <AccountDetailsPage 
        match={{params: { id: '123' }}}
        loadAccountDetails={loadAccountDetails}
        loading={false}
        accountDetails={{ AccountID: '123', Currency: 'USD', Amount: 1000, Type: 'Credit' }}
        history={history}
      />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })


  it('renders with props', () => {
    const page = mount(
      <AccountDetailsPage 
        deleteAccount={deleteAccount} match={{params: { id: '123' }}} 
        loadAccountDetails={loadAccountDetails}
        loading={false}
      />
    )
    expect(page).toBeDefined()
  })


  it('renders loading message when account details begin loading', () => {
    const tree = renderer.create(
      <AccountDetailsPage
        deleteAccount={deleteAccount}
        match={{ params: { id: '123' } }}
        loadAccountDetails={loadAccountDetails}
        loading={true}
      />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders bad account details message when account details cannot be loaded', () => {        
    const tree = renderer.create(
      <AccountDetailsPage 
        deleteAccount={deleteAccount}
        match={{params: { id: '123' }}}
        loadAccountDetails={loadAccountDetails}
        loading={false}
        accountDetails={undefined} 
      />
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('redirects to accounts page when account is deleted successfully', (done) => {
    const promise = {then: resolve => resolve()}
    const deleteAccount = () => promise

    const history = { push: jest.fn() }
    const page = shallow(
      <AccountDetailsPage
        deleteAccount={deleteAccount}
        match={{ params: { id: '123' } }}
        loadAccountDetails={loadAccountDetails}
        loading={false}
        accountDetails={{ AccountID: '123' }}
        history={history}
      />
    )

    const button = page.find('button').at(2)
    button.simulate('click')
    
    setTimeout(() => {
        try {
            expect(history.push).toBeCalledWith('/accounts')
            done()
        } catch(e) {
            done.fail(e)
        }
    }, 0)    
  })

  it('goes to account transaction page when user clicks on the account history button', () => {
    const history = { push: jest.fn() }
    const page = shallow(
      <AccountDetailsPage
        match={{ params: { id: '123' } }}
        loadAccountDetails={loadAccountDetails}
        loading={false}
        accountDetails={{ AccountID: '123' }}
        history={history}
      />
    )
    const button = page.find('button').at(3)
    button.simulate('click')
    expect(history.push).toBeCalledWith('/transactions/123')
  })

})