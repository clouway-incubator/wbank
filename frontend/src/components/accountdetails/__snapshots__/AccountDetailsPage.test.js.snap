// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountDetailsPage renders bad account details message when account details cannot be loaded 1`] = `
<p>
  Bad Account ID. Please go back.
</p>
`;

exports[`AccountDetailsPage renders loading message when account details begin loading 1`] = `
<div>
  Loading...
</div>
`;

exports[`AccountDetailsPage renders the account details properly 1`] = `
<div
  className="jumbotron"
>
  <h2>
    Account Details
  </h2>
  <table
    className="table"
  >
    <thead>
      <tr>
        <th>
          Account ID
        </th>
        <th>
          Currency
        </th>
        <th>
          Balance
        </th>
        <th>
          Type
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          123
        </td>
        <td>
          USD
        </td>
        <td>
          1000
        </td>
        <td>
          Credit
        </td>
      </tr>
    </tbody>
  </table>
  <button
    className="btn btn-default btn-md"
    onClick={[Function]}
  >
    Deposit
  </button>
  <button
    className="btn btn-default btn-md"
    onClick={[Function]}
  >
    Withdraw
  </button>
  <button
    className="btn btn-default btn-md"
    onClick={[Function]}
  >
    Delete
  </button>
  <button
    className="btn btn-default btn-md"
    onClick={[Function]}
  >
    Transactions History
  </button>
  <div
    className="div--back-button"
  >
    <button
      className="btn btn-default btn-md"
      onClick={[Function]}
    >
      Back
    </button>
  </div>
</div>
`;
